backend/api/serializers/computetask.py

```diff
@@ -59,9 +61,31 @@ class Meta:
             "identifier",
             "permissions",
             "transient",
+            "value",
         ]
 
     permissions = make_download_process_permission_serializer()(source="*")
+    value = serializers.SerializerMethodField(source="*", read_only=True)
+
+    def get_value(self, task_output):
+        data = []
+        for output_asset in task_output.assets.all():
+            if output_asset.asset_kind == ComputeTaskOutputAsset.Kind.ASSET_MODEL:
+                model = Model.objects.get(key=output_asset.asset_key)
```

Comment:
The `ComputeTaskOutputAsset` should be created after the `Model` as it need the model key.
I think the creation workflow is:
- ComputeTask
- ComputeTaskOutput
- Model/Performance
- ComputeTaskOutputAsset
Could you confirm @mblottiere?

Reply:
Yes, the orc will always send the output asset event after the referenced asset

Reply:
Perfect, thanks!

Reply:
One e2e [test](https://github.com/owkin/connect-ci/actions/runs/3068044196/jobs/4955020329) has raised an error at this level. So we have a bug somewhere.

Tests output:
```
    @pytest.mark.slow
    def test_federated_learning_workflow(factory, client, default_datasets):
        """Test federated learning workflow on each organization."""
    
        # create test environment
        spec = factory.create_algo(AlgoCategory.simple)
        algo = client.add_algo(spec)
    
        # create 1 traintuple per dataset and chain them
        traintuple = None
        rank = 0
        compute_plan_key = None
    
        # default_datasets contains datasets on each organization and
        # that has a result we can use for federated learning
        for dataset in default_datasets:
    
            traintuples = [traintuple.key] if traintuple else []
    
            spec = factory.create_traintuple(
                algo=algo,
                inputs=dataset.train_data_inputs + FLTaskInputGenerator.trains_to_train(traintuples),
                tag="foo",
                rank=rank,
                compute_plan_key=compute_plan_key,
            )
            traintuple = client.add_traintuple(spec)
>           traintuple = client.wait(traintuple)

url = 'http://backend-org-1-substra-backend-server.org-1.svc.cluster.local:8000/traintuple/14df92ab-7311-46f1-ac81-0e2c98c44e2b/'
```
Backend server output
```
GET /traintuple/14df92ab-7311-46f1-ac81-0e2c98c44e2b/

Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 197, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/viewsets.py", line 125, in view
    return self.dispatch(request, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 509, in dispatch
    response = self.handle_exception(exc)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 469, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception
    raise exc
  File "/usr/local/lib/python3.9/site-packages/rest_framework/views.py", line 506, in dispatch
    response = handler(request, *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/mixins.py", line 56, in retrieve
    return Response(serializer.data)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 555, in data
    ret = super().data
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 253, in data
    self._data = self.to_representation(self.instance)
  File "/usr/src/app/api/serializers/computetask.py", line 448, in to_representation
    data = super().to_representation(instance)
  File "/usr/src/app/api/serializers/computetask.py", line 277, in to_representation
    data = super().to_representation(instance)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 522, in to_representation
    ret[field.field_name] = field.to_representation(attribute)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 686, in to_representation
    return [
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 687, in <listcomp>
    self.child.to_representation(item) for item in iterable
  File "/usr/local/lib/python3.9/site-packages/rest_framework/serializers.py", line 522, in to_representation
    ret[field.field_name] = field.to_representation(attribute)
  File "/usr/local/lib/python3.9/site-packages/rest_framework/fields.py", line 1886, in to_representation
    return method(value)
  File "/usr/src/app/api/serializers/computetask.py", line 74, in get_value
    model = Model.objects.get(key=output_asset.asset_key)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 496, in get
    raise self.model.DoesNotExist(
api.models.model.Model.DoesNotExist: Model matching query does not exist.
```

Reply:
We use multiple threads for the sync, but only one thread per channel so I suppose this is not the issue.

Reply:
Can't this be just a race condition in the test ? The `traintuple = client.wait(traintuple)` guarantees the task is synced but not necessarily its output assets, does it ?

Reply:
Maybe there is flaw in the sync: we can fetch a task output from another channel and create an output asset upon it. I fix it in this [commit](https://github.com/Substra/substra-backend/pull/451/commits/0dddd6b70ba15cd98b664b7a6419f693c70a7f1a)

Reply:
In the next PR, I will add channel column in the input / output tables to help prevent this kind of issue

Reply:
Finally this should not be the source of the issue, because the task uuid is only for a specific channel so there is no risk to attach an output asset to a wrong task output.

Reply:
Maybe this is on the orchestrator side.

Reply:
Finally, it seems that the orchestrator does not create the events in the expected order:

```sql
SELECT event_kind, asset_kind, asset_key, timestamp FROM events;

-[ RECORD 28 ]-----------------------------------
event_kind | EVENT_ASSET_UPDATED
asset_kind | ASSET_COMPUTE_TASK
asset_key  | f4aff442-5513-4232-85ce-d043b983ed48
timestamp  | 2022-09-19 09:35:41.772171+00
-[ RECORD 29 ]-----------------------------------
event_kind | EVENT_ASSET_CREATED
asset_kind | ASSET_COMPUTE_TASK_OUTPUT_ASSET
asset_key  | f4aff442-5513-4232-85ce-d043b983ed48
timestamp  | 2022-09-19 09:36:15.178196+00
-[ RECORD 30 ]-----------------------------------
event_kind | EVENT_ASSET_CREATED
asset_kind | ASSET_MODEL
asset_key  | 7729e067-3521-4f45-9f3d-2ef51f3e66c3
timestamp  | 2022-09-19 09:36:15.178196+00
```

Reply:
> Yes, the orc will always send the output asset event after the referenced asset

🤦 it turns out that it did not, see https://github.com/Substra/orchestrator/pull/26

Reply:
thanks